File: backend/src/model/instance.py
from enum import StrEnum
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, Field

from model.world import WorldInfo
from model.group import GroupAccessType


class InstanceType(StrEnum):
    PUBLIC = "public"
    HIDDEN = "hidden"
    FRIENDS = "friends"
    PRIVATE = "private"
    GROUP = "group"


class Region(StrEnum):
    JP = "jp"
    USE = "use"
    USW = "us"
    EU = "eu"


class InstanceType(StrEnum):
    PUBLIC = "public"
    HIDDEN = "hidden"
    FRIENDS = "friends"
    PRIVATE = "private"
    GROUP = "group"


class ContentSettings(BaseModel):
    drones: Optional[bool] = Field(None)
    emoji: Optional[bool] = Field(None)
    pedestals: Optional[bool] = Field(None)
    prints: Optional[bool] = Field(None)
    stickers: Optional[bool] = Field(None)
    props: Optional[bool] = Field(None)


class CreateInstanceConfig(BaseModel):
    world_id: str = Field(..., alias="worldId")
    type: InstanceType = Field(..., alias="type")
    region: Optional[Region] = Field(None, alias="region")
    owner_id: Optional[str] = Field(None, alias="ownerId")
    role_ids: Optional[list[str]] = Field(None, alias="roleIds")
    group_access_type: Optional[GroupAccessType] = Field(None, alias="groupAccessType")
    queue_enabled: Optional[bool] = Field(None, alias="queueEnabled")
    display_name: Optional[str] = Field(None, alias="displayName")
    content_settings: Optional[ContentSettings] = Field(None, alias="contentSettings")


class InstanceInfo(BaseModel):
    id: str  # インスタンスID(wrld+instance)
    display_name: Optional[str] = Field(None, alias="displayName")  # 表示名
    name: str  # インスタンス番号
    location: str  # ロケーション
    type: InstanceType  # インスタンスの種類
    group_access_type: Optional[GroupAccessType] = Field(
        None, alias="groupAccessType"
    )  # グループインスタンスの種類
    instance_id: str = Field(..., alias="instanceId")  # インスタンスID(wrld+instance)
    short_name: str = Field(..., alias="secureName")  # 短縮名 (APIではsecureName,バグ?)
    user_count: int = Field(..., alias="userCount")  # 現在のユーザー数
    queue_enabled: bool = Field(..., alias="queueEnabled")  # キュー有効
    queue_size: int = Field(..., alias="queueSize")  # キューの人数
    region: str  # リージョン
    tags: list[str]  # タグ
    closed_at: Optional[datetime] = Field(None, alias="closedAt")
    world: WorldInfo  # ワールド情報
    world_id: str = Field(..., alias="worldId")  # ワールドID
    owner_id: Optional[str] = Field(
        None, alias="ownerId"
    )  # オーナーのID(Public -> None, Group -> groupId)

File: frontend/src/types/instance.ts
import type { GroupAccessType } from "./group";
import type { WorldInfo } from "./world";

export enum InstanceType {
    PUBLIC = "public",
    HIDDEN = "hidden",
    FRIENDS = "friends",
    PRIVATE = "private",
    GROUP = "group",
}

export enum Region {
    JP = "jp",
    USE = "use",
    USW = "us",
    EU = "eu",
}

export interface ContentSettings {
    drones?: boolean;
    emoji?: boolean;
    pedestals?: boolean;
    prints?: boolean;
    stickers?: boolean;
    props?: boolean;
}

export interface CreateInstanceConfig {
    worldId: string;
    type: InstanceType;
    region?: Region;
    ownerId?: string;
    roleIds?: string[];
    groupAccessType?: GroupAccessType;
    queueEnabled?: boolean;
    displayName?: string;
    contentSettings?: ContentSettings;
}

export interface InstanceInfo {
    id: string;                         // インスタンスID(wrld+instance)
    displayName?: string;               // 表示名
    name: string;                       // インスタンス番号
    location: string;                   // ロケーション
    type: InstanceType;                 // インスタンスの種類
    groupAccessType?: GroupAccessType;  // グループインスタンスの種類
    instanceId: string;                 // インスタンスID(wrld+instance)
    secureName: string;                 // 短縮名(APIではsecureName)
    userCount: number;                  // 現在のユーザー数
    queueEnabled: boolean;              // キュー有効
    queueSize: number;                  // キューの人数
    region: Region;                     // リージョン
    tags: string[];                     // タグ
    closedAt?: string;                  // datetime → ISO string
    world: WorldInfo;                   // ワールド情報
    worldId: string;                    // ワールドID
    ownerId?: string;                   // オーナーID
}
